pkg/sentry/fsimpl/fuse/fusefs.go

```diff
@@ -207,23 +210,163 @@ type Inode struct {
+func (fs *filesystem) newInode(creds *auth.Credentials, mode linux.FileMode) *dentry {
 	i := &Inode{}
 	i.InodeAttrs.Init(creds, linux.UNNAMED_MAJOR, fs.devMinor, fs.NextIno(), linux.ModeDirectory|0755)
 	i.OrderedChildren.Init(kernfs.OrderedChildrenOptions{})
-	i.dentry.Init(i)
+	i.dentry = fs.newDentry(i)
 
-	return &i.dentry
+	return i.dentry
 }
 
 // Open implements kernfs.Inode.Open.
 func (i *Inode) Open(ctx context.Context, rp *vfs.ResolvingPath, vfsd *vfs.Dentry, opts vfs.OpenOptions) (*vfs.FileDescription, error) {
-	fd, err := kernfs.NewGenericDirectoryFD(rp.Mount(), vfsd, &i.OrderedChildren, &i.locks, &opts)
-	if err != nil {
+	fuseConnection := i.dentry.fs.conn
+	if opts.Flags&linux.O_TRUNC != 0 && fuseConnection.AtomicOTrunc && fuseConnection.WritebackCache {
+		i.imutex.Lock()
+		defer i.imutex.Unlock()
+	}
+
+	if opts.Flags&linux.O_LARGEFILE == 0 && i.size > linux.MAX_NON_LFS {
+		return nil, syserror.EOVERFLOW
+	}
+
+	fd := fileDescription{}
+	out := linux.FUSEOpenOut{}
+	if !fuseConnection.NoOpen || opts.Mode.IsDir() {
+		kernelTask := kernel.TaskFromContext(ctx)
+		if kernelTask == nil {
+			log.Warningf("fusefs.DeviceFD.Read: couldn't get kernel task from context")
+			return nil, syserror.EINVAL
+		}
+
+		var opcode linux.FUSEOpcode
+		if opts.Mode.IsDir() {
+			opcode = linux.FUSE_OPENDIR
+		} else {
+			opcode = linux.FUSE_OPEN
+		}
+		in := linux.FUSEOpenIn{Flags: opts.Flags & ^uint32(linux.O_CREAT|linux.O_EXCL|linux.O_NOCTTY)}
+		if !fuseConnection.AtomicOTrunc {
+			in.Flags &= ^uint32(linux.O_TRUNC)
+		}
+		req, err := fuseConnection.NewRequest(auth.CredentialsFromContext(ctx), uint32(kernelTask.ThreadID()), i.Ino(), opcode, &in)
+		if err != nil {
+			return nil, err
+		}
+
+		res, err := fuseConnection.Call(kernelTask, req)
+		if err == syserror.ENOSYS && !opts.Mode.IsDir() {
+			fuseConnection.NoOpen = true
+		} else if err != nil {
+			return nil, err
+		}
+		if err := res.UnmarshalPayload(&out); err != nil {
+			return nil, err
+		}
+		fd.OpenFlag = out.OpenFlag
+	}
+	if opts.Mode.IsDir() {
+		fd.OpenFlag &= ^uint32(linux.FOPEN_DIRECT_IO)
+	}
+	fd.vfsfd.IncRef()
+	fd.Fh = out.Fh
+
+	// TODO(gvisor.dev/issue/3234): invalidate mmap after mmap had been implemented for FUSE Inode
+	fd.directIO = fd.OpenFlag&linux.FOPEN_DIRECT_IO != 0
+	fdOptions := &vfs.FileDescriptionOptions{}
+	if fd.OpenFlag&linux.FOPEN_NONSEEKABLE != 0 {
+		fdOptions.DenyPRead = true
+		fdOptions.DenyPWrite = true
+		fd.Nonseekable = true
+	}
+	if fuseConnection.AtomicOTrunc && opts.Flags&linux.O_TRUNC != 0 {
+		fuseConnection.Lock.Lock()
+		fuseConnection.AttributeVersion++
+		i.attributeVersion = fuseConnection.AttributeVersion
+		i.size = 0
+		fuseConnection.Lock.Unlock()
+		i.attributeTime = 0
+		if fuseConnection.WritebackCache {
+			i.ctime = ktime.NowFromContext(ctx)
+			i.mtime = ktime.NowFromContext(ctx)
+			i.version++
+		}
+	}
+
+	if err := fd.vfsfd.Init(&fd, opts.Flags, rp.Mount(), vfsd, fdOptions); err != nil {
```

Comment:
It seems that we could encapsulate this (vfsfd.Init() and the above fd := fileDescription{}) into a constructor for our file description.